File: linked-list/linked-list.cpp
#include <iostream>
using namespace std;

/* List node */
class Node
{
public:
	Node(int key);	/* Node constructor */
	int key;		/* Node value */
	Node *next;     /* Points to next Node */
};

Node::Node(int key) {
	this->key = key;
	this->next = nullptr;
}

/* Linked List (LL)*/
class LinkedList
{
public:
	LinkedList();				/* Linked List constructor */
	void push_back(int key)	;	/* Insert a Node at the end of the LL */
	void push_front(int key);	/* Insert a Node at the begin of the LL */
	void print();				/* DEBUG: print all LL elements */
private:
	int _size;					/* Number of elements in the LL	*/
	Node *head;					/* Points to the first LL element */
	Node *tail;					/* Points to the last LL element */
};

LinkedList::LinkedList() {
	this->_size = 0;
	this->head = nullptr;
}

void LinkedList::push_back(int key) {
	Node *node = new Node(key);   	// create a new Node MUST use 'new' keyword

	if (this->head == nullptr) {
		this->head = node; 			// point to new Node
		this->tail = node;
		this->_size += 1;
 	}
	else {
		this->tail->next = node;
		this->tail = node;
		this->_size += 1;
	}
}

void LinkedList::push_front(int key) {
	Node *node = new Node(key);   	// create a new Node MUST use 'new' keyword

	if (this->head == nullptr) {
		this->head = node; 			// point to new Node
		this->tail = node;
		this->_size += 1;
 	}
	else {
		node->next = this->head;
		this->head = node;
		this->_size += 1;
	}
}

void LinkedList::print() {
	Node *currentNodePtr = this->head;

	while (currentNodePtr != nullptr) {
		cout << "Element: " << currentNodePtr->key;
		cout << " Nexy Ptr: " <<  currentNodePtr->next << endl;
		currentNodePtr = currentNodePtr->next;
	}
}


int main() {
	LinkedList ll;

	ll.push_back(9);
	ll.push_back(8);
	ll.push_back(7);
	ll.push_back(6);
	ll.push_front(10);
	ll.push_front(11);

	ll.print();

	return 0;
}

File: linked-list/linked_list.py
class ListNode:
	"""Linked List node"""
	def __init__(self, value):
		'''Creata a new ListNode with the given value'''
		self._value = value
		self._next = None

	def __str__(self):
		return str(self._value) + "->" + str(self._next)

	def getValue(self):
		'''Retrurn the node value'''
		return self._value

	def setNext(self, listNode):
		'''Set next ListNode'''
		self._next = listNode

	def getNext(self):
		'''Return next ListNode'''
		return self._next

class LinkedList:
	"""Linked List Data Structure"""
	def __init__(self):
		'''Create a new empty Linked List'''
		self._head = None
		self._tail = None
		self._size = 0

	def __str__(self):
		return str(self._head) 
		
	def isEmpty(self):
		'''Return true if the Linked List is empty'''
		return self._size == 0
	def size(self):
		'''Return the number of elements'''
		return self._size

	def pushBack(self, value):
		'''Add a element at the end'''
		newNode = ListNode(value)
		if self.isEmpty():
			self._head = newNode
			self._tail = self._head
		else:
			self._tail.setNext(newNode)
			self._tail = self._tail.getNext()
		self._size += 1

	def popFront(self):
		'''
		Return and remove the first element

		Raise an Empty error if is empty
		'''
		if self.isEmpty():
			raise Empty('Linked List is empty')
		else:
			value = self._head.getValue()
			if self.size() == 1:
				self._head = None
				self._tail = None
			else:
				self._head = self._head.getNext() 
			self._size -= 1
		return value



		
		